measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+                                    # sum up all timepoints selections
+                                    sumi += m.cov_y[surro_idx, surro_idx]
+                                    count += 1
+
+                                return sumi <= 1
+
+                            con_name = "con_discreti_" + str(i) + str(t)
+                            m.add_component(con_name, pyo.Constraint(expr=discretizer))
+
+            # total cost
+            m.cost = pyo.Var(initialize=cost_initial)
+            # compute total cost
+            m.cost_compute = pyo.Constraint(rule=cost_compute)
+            # make total cost < budget
+            m.budget_limit = pyo.Constraint(rule=cost_limit)
+
+            # add model to object. This model lacks objective function which will be added separately later
+            self.mod = m
+
+    def _add_objective(
+        self,
+        obj=ObjectiveLib.A,
+        mix_obj=False,
+        alpha=1,
+    ):
+        """
+        Add objective function to the model.
+
+        This function is built for a consideration to easily add other objective functions
+        since this is a comparatively large code module.
+
+        Arguments
+        ---------
+        :param obj: Enum
+            "A" or "D" optimality, use trace or determinant of FIM
+        :param mix_obj: boolean
+            if True, the objective function is a weighted sum of A- and D-optimality (trace and determinant)
+        :param alpha: float
+            range [0,1], weight of mix_obj. if 1, it is A-optimality. if 0, it is D-optimality
         """
-        Compute FIM given a set of measurements
 
+        # set up design criterion
+        def compute_trace(m):
+            """compute trace
+            trace = sum(diag(M))
+            """
+            sum_x = sum(m.total_fim[j, j] for j in m.dim_fim)
+            return sum_x
+
+        # set objective
+        if obj == ObjectiveLib.A:  # A-optimailty
+            self.mod.Obj = pyo.Objective(rule=compute_trace, sense=pyo.maximize)
+
+        elif obj == ObjectiveLib.D:  # D-optimality
+
+            def _model_i(b):
+                # build grey-box module
+                self._build_model_external(b)
+
+            self.mod.my_block = pyo.Block(rule=_model_i)
+
+            # loop over parameters
+            for i in range(self.sens_info.n_parameters):
+                # loop over upper triangle of FIM
+                for j in range(i, self.sens_info.n_parameters):
+
+                    def eq_fim(m):
+                        """Make FIM in this model equal to the FIM computed by grey-box. Necessary."""
+                        return m.total_fim[i, j] == m.my_block.egb.inputs[(i, j)]
+
+                    con_name = "con" + str(i) + str(j)
+                    self.mod.add_component(con_name, pyo.Constraint(expr=eq_fim))
+
+            # add objective
+            # if mix_obj, we use a weighted sum of A- and D-optimality
+            if mix_obj:
```

Comment:
Why does mix_obj = True require obj == ObjectiveLib.D? Shouldn't obj not matter is mix_obj = True?

Reply:
Yes. fixed this logic